File: ruby-backend/app/controllers/auth_controller.rb
# Authentication controller - Laravel API compatibility
require 'json'
require_relative '../services/auth_service'

class AuthController
  class << self
    # Register a new user (POST /api/auth/register)
    def register(params)
      begin
        # Extract and validate parameters
        user_data = {
          username: params['username'],
          email: params['email'],
          password: params['password'],
          password_confirmation: params['password_confirmation']
        }
        
        # Register the user
        user = AuthService.register_user(user_data)
        
        # Generate JWT token
        token = AuthService.generate_jwt(user)
        
        # Return Laravel-compatible response
        {
          status: 201,
          body: {
            success: true,
            message: 'User registered successfully',
            data: {
              user: {
                id: user.id,
                username: user.username,
                email: user.email,
                created_at: user.created_at&.iso8601
              },
              token: token,
              token_type: 'bearer',
              expires_in: SpotikConfig::Settings.jwt_ttl * 60 # Convert minutes to seconds
            }
          }
        }
        
      rescue ValidationError => e
        {
          status: 422,
          body: {
            success: false,
            message: 'Validation failed',
            errors: e.errors
          }
        }
      rescue => e
        {
          status: 500,
          body: {
            success: false,
            message: 'Registration failed',
            error: e.message
          }
        }
      end
    end
    
    # Authenticate user and return JWT token (POST /api/auth/login)
    def login(params)
      begin
        # Validate required parameters
        if params['email'].nil? || params['email'].strip.empty?
          return {
            status: 422,
            body: {
              success: false,
              message: 'Validation failed',
              errors: { email: ['The email field is required.'] }
            }
          }
        end
        
        if params['password'].nil? || params['password'].empty?
          return {
            status: 422,
            body: {
              success: false,
              message: 'Validation failed',
              errors: { password: ['The password field is required.'] }
            }
          }
        end
        
        # Validate email format
        unless params['email'].match?(/\A[^@\s]+@[^@\s]+\z/)
          return {
            status: 422,
            body: {
              success: false,
              message: 'Validation failed',
              errors: { email: ['The email must be a valid email address.'] }
            }
          }
        end
        
        # Authenticate user
        user = AuthService.authenticate(params['email'], params['password'])
        
        unless user
          return {
            status: 401,
            body: {
              success: false,
              message: 'Invalid credentials',
              error: 'The provided credentials are incorrect.'
            }
          }
        end
        
        # Generate JWT token
        token = AuthService.generate_jwt(user)
        
        # Return Laravel-compatible response
        {
          status: 200,
          body: {
            success: true,
            message: 'Login successful',
            data: {
              user: {
                id: user.id,
                username: user.username,
                email: user.email,
                created_at: user.created_at&.iso8601
              },
              token: token,
              token_type: 'bearer',
              expires_in: SpotikConfig::Settings.jwt_ttl * 60 # Convert minutes to seconds
            }
          }
        }
        
      rescue => e
        {
          status: 500,
          body: {
            success: false,
            message: 'Login failed',
            error: e.message
          }
        }
      end
    end
    
    # Get the authenticated user (GET /api/auth/me)
    def me(token)
      begin
        auth_data = AuthService.validate_jwt(token)
        
        unless auth_data
          return {
            status: 401,
            body: {
              success: false,
              message: 'Token not provided or invalid',
              error: 'Token absent or invalid'
            }
          }
        end
        
        user = auth_data[:user]
        
        {
          status: 200,
          body: {
            success: true,
            message: 'User retrieved successfully',
            data: {
              user: {
                id: user.id,
                username: user.username,
                email: user.email,
                created_at: user.created_at&.iso8601,
                updated_at: user.updated_at&.iso8601
              }
            }
          }
        }
        
      rescue AuthenticationError => e
        case e.code
        when :token_expired
          {
            status: 401,
            body: {
              success: false,
              message: 'Token has expired',
              error: 'Token expired'
            }
          }
        when :token_invalid
          {
            status: 401,
            body: {
              success: false,
              message: 'Token is invalid',
              error: 'Invalid token'
            }
          }
        when :token_absent
          {
            status: 401,
            body: {
              success: false,
              message: 'Token not provided',
              error: 'Token absent'
            }
          }
        else
          {
            status: 401,
            body: {
              success: false,
              message: 'Authentication failed',
              error: e.message
            }
          }
        end
      rescue => e
        {
          status: 500,
          body: {
            success: false,
            message: 'Could not retrieve user',
            error: e.message
          }
        }
      end
    end
    
    # Refresh the JWT token (POST /api/auth/refresh)
    def refresh(token)
      begin
        new_token = AuthService.refresh_jwt(token)
        
        unless new_token
          return {
            status: 401,
            body: {
              success: false,
              message: 'Token not provided or invalid',
              error: 'Token absent or invalid'
            }
          }
        end
        
        {
          status: 200,
          body: {
            success: true,
            message: 'Token refreshed successfully',
            data: {
              token: new_token,
              token_type: 'bearer',
              expires_in: SpotikConfig::Settings.jwt_ttl * 60 # Convert minutes to seconds
            }
          }
        }
        
      rescue AuthenticationError => e
        case e.code
        when :token_expired
          {
            status: 401,
            body: {
              success: false,
              message: 'Token has expired and cannot be refreshed',
              error: 'Token expired'
            }
          }
        when :token_invalid
          {
            status: 401,
            body: {
              success: false,
              message: 'Token is invalid',
              error: 'Invalid token'
            }
          }
        else
          {
            status: 401,
            body: {
              success: false,
              message: 'Token not provided',
              error: 'Token absent'
            }
          }
        end
      rescue => e
        {
          status: 500,
          body: {
            success: false,
            message: 'Could not refresh token',
            error: e.message
          }
        }
      end
    end
    
    # Logout user and invalidate token (POST /api/auth/logout)
    def logout(token)
      begin
        # Note: In a production system, you would want to implement token blacklisting
        # For now, we'll just return success (client should discard the token)
        
        {
          status: 200,
          body: {
            success: true,
            message: 'Successfully logged out'
          }
        }
        
      rescue => e
        # Even if there's an error, return success for logout
        # This allows logout to work even without a valid token
        {
          status: 200,
          body: {
            success: true,
            message: 'Successfully logged out'
          }
        }
      end
    end
  end
end

File: backend/app/Http/Controllers/AuthController.php
<?php

namespace App\Http\Controllers;

use App\Models\User;
use Illuminate\Http\Request;
use Illuminate\Http\JsonResponse;
use Illuminate\Support\Facades\Hash;
use Illuminate\Support\Facades\Validator;
use Illuminate\Validation\ValidationException;
use Tymon\JWTAuth\Facades\JWTAuth;
use Tymon\JWTAuth\Exceptions\JWTException;
use Tymon\JWTAuth\Exceptions\TokenExpiredException;
use Tymon\JWTAuth\Exceptions\TokenInvalidException;

class AuthController extends Controller
{
    /**
     * Register a new user
     */
    public function register(Request $request): JsonResponse
    {
        try {
            // Validate the request data
            $validator = Validator::make($request->all(), User::validationRules());

            if ($validator->fails()) {
                return response()->json([
                    'success' => false,
                    'message' => 'Validation failed',
                    'errors' => $validator->errors()
                ], 422);
            }

            // Create the user
            $user = User::create([
                'username' => $request->username,
                'email' => $request->email,
                'password_hash' => Hash::make($request->password),
            ]);

            // Generate JWT token for the new user
            $token = JWTAuth::fromUser($user);

            return response()->json([
                'success' => true,
                'message' => 'User registered successfully',
                'data' => [
                    'user' => [
                        'id' => $user->id,
                        'username' => $user->username,
                        'email' => $user->email,
                        'created_at' => $user->created_at,
                    ],
                    'token' => $token,
                    'token_type' => 'bearer',
                    'expires_in' => config('jwt.ttl') * 60, // Convert minutes to seconds
                ]
            ], 201);

        } catch (\Exception $e) {
            return response()->json([
                'success' => false,
                'message' => 'Registration failed',
                'error' => $e->getMessage()
            ], 500);
        }
    }

    /**
     * Authenticate user and return JWT token
     */
    public function login(Request $request): JsonResponse
    {
        try {
            // Validate the request data
            $validator = Validator::make($request->all(), [
                'email' => 'required|email',
                'password' => 'required|string|min:1',
            ]);

            if ($validator->fails()) {
                return response()->json([
                    'success' => false,
                    'message' => 'Validation failed',
                    'errors' => $validator->errors()
                ], 422);
            }

            // Find user by email
            $user = User::where('email', $request->email)->first();

            if (!$user) {
                return response()->json([
                    'success' => false,
                    'message' => 'Invalid credentials',
                    'error' => 'User not found'
                ], 401);
            }

            // Verify password
            if (!Hash::check($request->password, $user->password_hash)) {
                return response()->json([
                    'success' => false,
                    'message' => 'Invalid credentials',
                    'error' => 'Password does not match'
                ], 401);
            }

            // Generate JWT token
            $token = JWTAuth::fromUser($user);

            return response()->json([
                'success' => true,
                'message' => 'Login successful',
                'data' => [
                    'user' => [
                        'id' => $user->id,
                        'username' => $user->username,
                        'email' => $user->email,
                        'created_at' => $user->created_at,
                    ],
                    'token' => $token,
                    'token_type' => 'bearer',
                    'expires_in' => config('jwt.ttl') * 60, // Convert minutes to seconds
                ]
            ], 200);

        } catch (JWTException $e) {
            return response()->json([
                'success' => false,
                'message' => 'Could not create token',
                'error' => $e->getMessage()
            ], 500);
        } catch (\Exception $e) {
            return response()->json([
                'success' => false,
                'message' => 'Login failed',
                'error' => $e->getMessage()
            ], 500);
        }
    }

    /**
     * Get the authenticated user
     */
    public function me(Request $request): JsonResponse
    {
        try {
            $user = JWTAuth::parseToken()->authenticate();

            if (!$user) {
                return response()->json([
                    'success' => false,
                    'message' => 'User not found'
                ], 404);
            }

            return response()->json([
                'success' => true,
                'message' => 'User retrieved successfully',
                'data' => [
                    'user' => [
                        'id' => $user->id,
                        'username' => $user->username,
                        'email' => $user->email,
                        'created_at' => $user->created_at,
                        'updated_at' => $user->updated_at,
                    ]
                ]
            ], 200);

        } catch (TokenExpiredException $e) {
            return response()->json([
                'success' => false,
                'message' => 'Token has expired',
                'error' => 'Token expired'
            ], 401);
        } catch (TokenInvalidException $e) {
            return response()->json([
                'success' => false,
                'message' => 'Token is invalid',
                'error' => 'Invalid token'
            ], 401);
        } catch (JWTException $e) {
            return response()->json([
                'success' => false,
                'message' => 'Token not provided',
                'error' => 'Token absent'
            ], 401);
        } catch (\Exception $e) {
            return response()->json([
                'success' => false,
                'message' => 'Could not retrieve user',
                'error' => $e->getMessage()
            ], 500);
        }
    }

    /**
     * Refresh the JWT token
     */
    public function refresh(Request $request): JsonResponse
    {
        try {
            $token = JWTAuth::parseToken()->refresh();

            return response()->json([
                'success' => true,
                'message' => 'Token refreshed successfully',
                'data' => [
                    'token' => $token,
                    'token_type' => 'bearer',
                    'expires_in' => config('jwt.ttl') * 60, // Convert minutes to seconds
                ]
            ], 200);

        } catch (TokenExpiredException $e) {
            return response()->json([
                'success' => false,
                'message' => 'Token has expired and cannot be refreshed',
                'error' => 'Token expired'
            ], 401);
        } catch (TokenInvalidException $e) {
            return response()->json([
                'success' => false,
                'message' => 'Token is invalid',
                'error' => 'Invalid token'
            ], 401);
        } catch (JWTException $e) {
            return response()->json([
                'success' => false,
                'message' => 'Token not provided',
                'error' => 'Token absent'
            ], 401);
        } catch (\Exception $e) {
            return response()->json([
                'success' => false,
                'message' => 'Could not refresh token',
                'error' => $e->getMessage()
            ], 500);
        }
    }

    /**
     * Logout user and invalidate token
     */
    public function logout(Request $request): JsonResponse
    {
        try {
            // Try to get the token, but don't fail if it's not present
            $token = JWTAuth::getToken();
            
            if ($token) {
                JWTAuth::invalidate($token);
            }

            return response()->json([
                'success' => true,
                'message' => 'Successfully logged out'
            ], 200);

        } catch (\Exception $e) {
            // Even if token invalidation fails, return success
            // This allows logout to work even without a valid token
            return response()->json([
                'success' => true,
                'message' => 'Successfully logged out'
            ], 200);
        }
    }
}